aten/src/ATen/native/cpu/DistanceOpsKernel.cpp

```diff
@@ -0,0 +1,192 @@
+#include "ATen/native/cpu/ReduceOpsKernel.h"
+
+#include <numeric>
+#include <iterator>
+#include <algorithm>
+
+#include "ATen/Dispatch.h"
+#include "ATen/Parallel.h"
+#include "ATen/core/optional.h"
+#include "ATen/cpu/vec256/vec256.h"
+
+namespace at { namespace native { namespace {
+
+template<typename scalar_t>
+struct PDist {
+
+  static inline scalar_t sign(scalar_t val) {
+    return (0 < val) - (val < 0);
+  }
+
+  static scalar_t zdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += *a != *b;
+    }
+    return result;
+  }
+
+  static scalar_t odist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += std::abs(*a - *b);
+    }
+    return result;
+  }
+
+  static scalar_t tdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      scalar_t diff = *a - *b;
+      result += diff * diff;
+    }
+    return std::sqrt(result);
+  }
+
+  static scalar_t pdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += std::pow(std::abs(*a - *b), p);
+    }
+    return std::pow(result, 1.0 / p);
+  }
+
+  static scalar_t idist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
```

Comment:
this gives the same answer for inf and -inf -- is that right?  It's a little unclear to me because the scipy docs say p >= 1, but it accepts other values...

Reply:
Yeah. I need to add a check that p must be in [0, inf]